lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+
+  auto sharedOrder = expandMatrixOrderWithBatch(sharedLayout.getOrder());
+  // compute contiguity of fastest dimension in shared layout.
+  unsigned vectorSize = sizePerThread[sharedOrder[0]];
+  if (sharedLayout.getMaxPhase() > 1)
+    vectorSize = std::min(vectorSize, sharedLayout.getVec());
+  auto vecTy = vec_ty(elemTy, vectorSize);
+  // loop increments depend on fastest dim
+  unsigned dimStep[3] = {1, 1, 1};
+  dimStep[sharedOrder[0]] = vectorSize;
+
+  auto shapePerCTABTile = shapePerCTATile[dim.batch];
+  auto shapePerCTANonKTile = shapePerCTATile[dim.nonK];
+  auto sizeBPerThread = sizePerThread[dim.batch];
+  auto sizeNonKPerThread = sizePerThread[dim.nonK];
+  auto numBTiles = std::max(1u, B / shapePerCTABTile);
+  auto numNonKTiles = std::max(1u, NonK / shapePerCTANonKTile);
+
+  auto perThreadShape =
+      getElemsPerThreadInOp(opTensorShape, shapePerCTATile, sizePerThread);
+
+  SmallVector<Value> opValues(numBTiles * sizeBPerThread * K * numNonKTiles *
+                              sizeNonKPerThread);
+
+  bool swizzlePath = isSwizzled(sharedLayout);
+
+  // In swizzled memory case basePtr stores pointer to the beginning of shared
+  // memmory object.
+  //
+  // If memory is not swizzled, algorithm breaks element offset pointer into
+  // constant and non-constant part. Non-constant (depends on thread id) part is
+  // same for all elements, so it is computed only once. basePtr stores this
+  // non-constant part
+  Value basePtr;
+  if (swizzlePath) {
+    basePtr = smem.base;
+  } else {
+    auto laneOffset = getUnswizzledFirstElemOffset(
+        rewriter, loc, B, NonK, bTileOffset, nonKTileOffset, strides[dim.batch],
+        strides[dim.nonK]);
+    basePtr = gep(ptrTy, elemTy, smem.base, laneOffset);
   }
-  auto elemTy = typeConverter->convertType(bTensorTy.getElementType());
-
-  Type ptrTy = bSmem.base.getType();
-  SmallVector<Value> bPtrs(bNumPtr);
-  for (int i = 0; i < bNumPtr; ++i)
-    bPtrs[i] = gep(ptrTy, elemTy, bSmem.base, bOff[i]);
-
-  SmallVector<Value> vbs;
-
-  int nShapePerCTATile = getShapePerCTATileForMN(dLayout, false /*isM*/);
-  int nSizePerThread = getSizePerThreadForMN(dLayout, false /*isM*/);
-
-  for (unsigned k = 0; k < K; ++k)
-    for (unsigned n = 0; n < N; n += nShapePerCTATile)
-      for (unsigned nn = 0; nn < nSizePerThread; ++nn) {
-        Value offset =
-            add(mul(i32_val(n + nn), strideBN), mul(i32_val(k), strideBK));
-        Value pb = gep(ptrTy, elemTy, bPtrs[0], offset);
-        Value vb = load(elemTy, pb);
-        vbs.emplace_back(vb);
-      }
 
-  return getStructFromValueTable(vbs, rewriter, loc, typeConverter, elemTy);
+  // This loop nest iterates over all values loaded in one thread across batch,
+  // k and nonK dimensions. Blocked dot operand layout allocates data in tiles
+  // of size <sizePerThread>*<threadsPerWarp>*<numberWarps> for batch and nonK
+  // dimensions, if tensor shape is larger than tile, pattern repeats. To take
```

Comment:
"dimensions, if tensor shape is larger.." --> "dimensions. If tensor shape is larger..."